### ISSUE TITLE
How to use Greedy Filter Pruning for MobileNetV1?

### ISSUE THREAD

**Issue by Bowenwu1** [NervanaSystems/distiller#282]
Hello, I want to employ the Greedy Filter Pruning Algorithm for MobileNetV1, but I find your code only support ResNet18, ResNet56, ResNet50. Then how to deal with MobileNetV1? Need I add extra code here?https://github.com/NervanaSystems/distiller/blob/3cde6c5eb68ed8e6e279b24f4d0283d7caed99ce/distiller/pruning/greedy_filter_pruning.py#L225-L259

**Comment by nzmora** [comment]
Hi @Bowenwu1,

Indeed, the implementation of the Greedy filter pruning algorithm is not generic and supports only a subset of the models.  I found this method to be prohibitively slow when using deep models with a large dataset such as ImageNet.
Adding mobilenet is pretty straight-forward, except that it uses depth-wise separable convolutions, which require some changes in the pruning code.  I have done this for a similar task, but I hesitate to add support for mobilenet, because I don't think this algorithm is suited to this size networks.

If you think otherwise, please let me know.  Perhaps I'm missing some usage.
Cheers,
Neta

**Comment by Bowenwu1** [comment]
```
mobilenetv1_params = ["module.layers.0.conv2.weight", "module.layers.1.conv2.weight", "module.layers.2.conv2.weight",
                      "module.layers.3.conv2.weight", "module.layers.4.conv2.weight", "module.layers.5.conv2.weight",
                      "module.layers.6.conv2.weight", "module.layers.7.conv2.weight", "module.layers.8.conv2.weight",

                      "module.layers.9.conv2.weight", "module.layers.10.conv2.weight", "module.layers.11.conv2.weight",
                      "module.layers.12.conv2.weight"]
```
First I add above code to greedy filter pruner. Then, with `fraction_to_prun=0.7` and `pruning_step=0.1, ` greedy filter pruning works fine for MobileNetV1 on Cifar10.
By the way, I'm confused about `because I don't think this algorithm is suited to this size networks` that you talk above.

**Comment by nzmora** [comment]
Hi @Bowenwu1,

I'm glad to hear that this algorithm works for you with MobileNet.  I'm surprised that it does, because I haven't taken care of the case of depth-wise convolutions.  

When I said that _"I don't think this algorithm is suited to this size networks"_ I meant that because Greedy needs to take a lot of steps, and between steps we need to fine-tune, this is not a "sample efficient" search algorithm.  MobileNet is usually used with ImageNet, which is very large (with large images), so not being sample-efficient means that it will run for a very long time.
With Cifar10 fine-tuning (training) is much faster so the picture is different.
Cheers
Neta

**Comment by Bowenwu1** [comment]
Thank you for your reply! @nzmora 
I plan to use greedy filter pruning for ImageNet case with finetuning with part of training data(eg. 10%), and I wish it will work OK.
Then, you actually support thining of depth-wise convolutions(group convolution) on [commit 6b832025](https://github.com/NervanaSystems/distiller/commit/6b8320253848ba0fed438c520e3b14e6008e4450) 
, so once some filters of conv2 are masked, the following layers will be correctly thining.
Furthermore, I find a bug of greedy filter pruner:
https://github.com/NervanaSystems/distiller/blob/bf111e4e80ff2f09f8d3653e7291e91e02f05f71/distiller/pruning/greedy_filter_pruning.py#L302-L309
After del zeros_mask_dict is called, you use zeros_mask_dict again and error arose.
And I open a PR for fixing this bug.[PR](https://github.com/NervanaSystems/distiller/pull/286)

**Comment by nzmora** [comment]
Excellent @Bowenwu1!  If you can share your results once you're, I'll be very happy to see them.
I will share a working implementation of MobileNet channel compression using Automated Model Compression in a couple of weeks, and comparing the results will be interesting.
Cheers,
Neta

**Comment by Bowenwu1** [comment]
Glad to hear that! @nzmora 
I'm planning to conduct AMC on MobileNet too. I will share my results later. 
Cheers,
Bob
